usim/_primitives/context.py

```diff
@@ -115,36 +138,77 @@ def do(
                 await containing_scope
                 print('... scope has finished')
 
-            with Scope() as scope:
+            async with Scope() as scope:
                 scope.do(graceful(scope))
 
-        All `volatile` activities are aborted at the end of the scope,
-        after all non-`volatile` activities have finished.
-        Aborting ``volatile` activities is not graceful:
+        All ``volatile`` activities are aborted at the end of the scope,
+        after all non-``volatile`` activities have finished.
+        Aborting ``volatile`` activities is not graceful:
         :py:class:`GeneratorExit` is raised in the activity,
-        which must exit without `await`\ ing or `yield`\ ing anything.
+        which must exit without ``await``\ ing or ``yield``\ ing anything.
         """
-        child_task = Task(payload)
+        assert after is None or at is None,\
+            "schedule date must be either absolute or relative"
+        assert after is None or after > 0,\
+            "schedule date must not be in the past"
+        assert at is None or at > __LOOP_STATE__.LOOP.time,\
+            "schedule date must not be in the past"
+        child_task = Task(payload, self, delay=after, at=at)
         __LOOP_STATE__.LOOP.schedule(
             child_task.__runner__,
-            delay=after, at=at
         )
         if not volatile:
             self._children.append(child_task)
         else:
             self._volatile_children.append(child_task)
         return child_task
 
+    def __cancel__(self):
+        """Cancel this scope"""
+        if self._interruptable:
+            __LOOP_STATE__.LOOP.schedule(self._activity, self._cancel_self)
+
+    def _disable_interrupts(self):
+        self._interruptable = False
+        self._cancel_self.revoke()
+
     async def _await_children(self):
         for child in self._children:
             await child.done
 
-    def _cancel_children(self):
+    def _reraise_concurrent(self):
+        """re-``raise`` any exceptions that occurred concurrently"""
+        suppress = self.SUPPRESS_CONCURRENT
+        promote = self.PROMOTE_CONCURRENT
+        concurrent = []
         for child in self._children:
-            child.cancel(self)
+            if child.status is TaskState.FAILED:
+                exc = child.__exception__
+                if isinstance(exc, promote):
+                    raise exc
+                if not isinstance(exc, suppress):
+                    concurrent.append(exc)
+        if concurrent:
+            raise Concurrent(*concurrent)
+
+    def _reraise_privileged(self):
```

Comment:
Please consider refactoring `_reraise_privileged` and `_reraise_concurrent` as the two methods are practically identically except the check for concurrent exceptions.